LIPS/lip-25.md

```diff
@@ -0,0 +1,1983 @@
+
+    cacheItem.stakingModuleAddress = stakingModuleData.stakingModuleAddress;
+    cacheItem.stakingModuleId = stakingModuleData.id;
+    cacheItem.stakingModuleFee = stakingModuleData.stakingModuleFee;
+    cacheItem.treasuryFee = stakingModuleData.treasuryFee;
+    cacheItem.stakeShareLimit = stakingModuleData.stakeShareLimit;
+    cacheItem.status = StakingModuleStatus(stakingModuleData.status);
+
+    (
+        uint256 totalExitedValidators,
+        uint256 totalDepositedValidators,
+        uint256 depositableValidatorsCount
+    ) = IStakingModule(cacheItem.stakingModuleAddress).getStakingModuleSummary();
+
+    cacheItem.availableValidatorsCount = cacheItem.status == StakingModuleStatus.Active
+        ? depositableValidatorsCount
+        : 0;
+
+    // the module might not receive all exited validators data yet => we need to replacing
+    // the exitedValidatorsCount with the one that the staking router is aware of
+    cacheItem.activeValidatorsCount =
+        totalDepositedValidators -
+        Math256.max(totalExitedValidators, stakingModuleData.exitedValidatorsCount);
+}
+
+function _getDepositsAllocation(
+    uint256 _depositsToAllocate
+)
+    internal
+    view
+    returns (uint256 allocated, uint256[] memory allocations, StakingModuleCache[] memory stakingModulesCache)
+{
+    // calculate total used validators for operators
+    uint256 totalActiveValidators;
+
+    (totalActiveValidators, stakingModulesCache) = _loadStakingModulesCache();
+
+    uint256 stakingModulesCount = stakingModulesCache.length;
+    allocations = new uint256[](stakingModulesCount);
+    if (stakingModulesCount > 0) {
+        /// @dev new estimated active validators count
+        totalActiveValidators += _depositsToAllocate;
+        uint256[] memory capacities = new uint256[](stakingModulesCount);
+        uint256 targetValidators;
+
+        for (uint256 i; i < stakingModulesCount; ) {
+            allocations[i] = stakingModulesCache[i].activeValidatorsCount;
+            targetValidators = (stakingModulesCache[i].stakeShareLimit * totalActiveValidators) / TOTAL_BASIS_POINTS;
+            capacities[i] = Math256.min(
+                targetValidators,
+                stakingModulesCache[i].activeValidatorsCount + stakingModulesCache[i].availableValidatorsCount
+        );
+        unchecked {
+            ++i;
+        }
+      }
+
+      (allocated, allocations) = MinFirstAllocationStrategy.allocate(allocations, capacities, _depositsToAllocate);
+    }
+  }
+```
+
+#### 2.2. Contract version upgrade
+
+For correct migration to the new version of the `StakingRouter` contract, the existing `initialize` external method should be updated, and the new `finalizeUpgrade_v2` external method should be implemented. The existing `ZeroAddress` error type should be transformed to the new `ZeroAddressLido` error type. Also, the new `ZeroAddressAdmin` error type should be added.
```

Comment:
> The existing `ZeroAddress` error type should be transformed to the new `ZeroAddressLido` error type. Also, the new `ZeroAddressAdmin` error type should be added.

It's bytecode optimisations, not sure it should be mentioned in the LIP

Reply:
OK, removed mentions of these errors from the 2.2 section.